ports/nrf/boards/feather52/board.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+#include <stdbool.h>
+
+#include "nrf.h"
+
+#include "boards/board.h"
+
+#if 0
+#include "common-hal/microcontroller/Pin.h"
+#include "hal/include/hal_gpio.h"
+#include "shared-bindings/digitalio/DigitalInOut.h"
+#include "shared-bindings/neopixel_write/__init__.h"
+#endif
+
+// Must match temp register in bootloader
+#define BOOTLOADER_VERSION_REGISTER     NRF_TIMER2->CC[0]
+uint32_t bootloaderVersion = 0;
+
+volatile uint32_t ticks_ms = 0;
+
+#define HAL_LFCLK_FREQ (32768UL)
+#define HAL_RTC_FREQ   (1024UL)
+#define HAL_RTC_COUNTER_PRESCALER ((HAL_LFCLK_FREQ/HAL_RTC_FREQ)-1)
+
+/* Maximum RTC ticks */
+#define portNRF_RTC_MAXTICKS   ((1U<<24)-1U)
+
+void board_init(void)
+{
+  // Retrieve bootloader version
+  bootloaderVersion = BOOTLOADER_VERSION_REGISTER;
+
+  // 32Khz XTAL
+  NRF_CLOCK->LFCLKSRC = (uint32_t)((CLOCK_LFCLKSRC_SRC_Xtal << CLOCK_LFCLKSRC_SRC_Pos) & CLOCK_LFCLKSRC_SRC_Msk);
+  NRF_CLOCK->TASKS_LFCLKSTART = 1UL;
+
+  // Set up RTC1 as tick timer
+  NVIC_DisableIRQ(RTC1_IRQn);
+  NRF_RTC1->EVTENCLR    = RTC_EVTEN_COMPARE0_Msk;
+  NRF_RTC1->INTENCLR    = RTC_INTENSET_COMPARE0_Msk;
+  NRF_RTC1->TASKS_STOP  = 1;
+  NRF_RTC1->TASKS_CLEAR = 1;
+
+  ticks_ms = 0;
+
+  NRF_RTC1->PRESCALER = HAL_RTC_COUNTER_PRESCALER;
+  NRF_RTC1->INTENSET = RTC_INTENSET_TICK_Msk;
+  NRF_RTC1->TASKS_START = 1;
+  NRF_RTC1->EVTENSET = RTC_EVTEN_OVRFLW_Msk;
+  NVIC_SetPriority(RTC1_IRQn, 0xf); // lowest priority
+  NVIC_EnableIRQ(RTC1_IRQn);
```

Comment:
In the atmel-samd port we handle this in the port level init and any high level clock config is done through defines. Are the existing boards different enough to warrant this done per board?

Additionally, should we remove the other dev boards we don't plan on supporting?

Reply:
I think we should remove other boards that we definitely won't be able to support, but I'd like to define what that list should be first so have kept things as is for the moment, and the files already exist in master so I think we can kick that can down the road to the next PR. They should go if we can't support them, though, especially with the API level changes required for maintenance.

Clock support on the nRF series does actually vary a lot board to board depending on HW design since there are several valid clock sources and configs. Some of our boards have a 32kHz XTAL for low power mode, some don't, etc., and I think clock setup with a low power chip like this generally is very board specific. That can be moved to macros, but I'm not convinced there's any big gain doing that except having to jump between files???

Reply:
I guess I'm thinking towards a world where we have multiple board designs of our own with nrf52 just like the M0. Its fine with me to wait until we actually need it refactored.